scanners/sslyze.py

```diff
@@ -312,12 +314,133 @@ def cert_issuer_name(parsed):
+        return None, None
+    except Exception as err:
+        utils.notify(err)
+        logging.warn("\tUnknown exception when initializing server connectivity info.")
+        return None, None
+
+    try:
+        logging.debug("\tTesting connectivity with timeout of %is." % network_timeout)
+        server_info.test_connectivity_to_server(network_timeout=network_timeout)
+    except sslyze.server_connectivity.ServerConnectivityError as err:
+        logging.warn("\tServer connectivity not established during test.")
+        return None, None
+    except Exception as err:
+        utils.notify(err)
+        logging.warn("\tUnknown exception when performing server connectivity info.")
+        return None, None
+
+    if sync:
+        scanner = SynchronousScanner()
+    else:
+        scanner = ConcurrentScanner()
+
+    return server_info, scanner
+
+
+# Run each scan in-process, one at a time.
+# Takes longer, but no multi-process funny business.
+def scan_serial(scanner, server_info, options):
+    logging.debug("\tRunning scans in serial.")
+    logging.debug("\t\tSSLv2 scan.")
+    sslv2 = scanner.run_scan_command(server_info, Sslv20ScanCommand())
+    logging.debug("\t\tSSLv3 scan.")
+    sslv3 = scanner.run_scan_command(server_info, Sslv30ScanCommand())
+    logging.debug("\t\tTLSv1.0 scan.")
+    tlsv1 = scanner.run_scan_command(server_info, Tlsv10ScanCommand())
+    logging.debug("\t\tTLSv1.1 scan.")
+    tlsv1_1 = scanner.run_scan_command(server_info, Tlsv11ScanCommand())
+    logging.debug("\t\tTLSv1.2 scan.")
+    tlsv1_2 = scanner.run_scan_command(server_info, Tlsv12ScanCommand())
+
+    # Default to cert info on
+    if options.get("sslyze-no-certs", False) is False:
+        logging.debug("\t\tCertificate information scan.")
+        certs = scanner.run_scan_command(server_info, CertificateInfoScanCommand())
     else:
-        return (len(target[protocol]["accepted_cipher_list"]) > 0)
+        certs = None
+
+    logging.debug("\tDone scanning.")
+
+    return sslv2, sslv3, tlsv1, tlsv1_1, tlsv1_2, certs
+
+
+# Run each scan in parallel, using multi-processing.
+# Faster, but can generate many processes.
+def scan_parallel(scanner, server_info, options):
+    logging.debug("\tRunning scans in parallel.")
+
+    def queue(command):
+        try:
+            return scanner.queue_scan_command(server_info, command)
+        except Exception as err:
+            utils.notify(err)
+            logging.warn("Unknown exception queueing sslyze command.")
+            return None
+
+    # Initialize commands and result containers
+    sslv2, sslv3, tlsv1, tlsv1_1, tlsv1_2, certs = None, None, None, None, None, None
+
+    # Queue them all up
+    queue(Sslv20ScanCommand())
+    queue(Sslv30ScanCommand())
+    queue(Tlsv10ScanCommand())
+    queue(Tlsv11ScanCommand())
+    queue(Tlsv12ScanCommand())
+
+    # Default to cert info on.
+    if options.get("sslyze-no-certs", False) is False:
+        queue(CertificateInfoScanCommand())
```

Comment:
The `--sslyze-no-certs` option is similarly undocumented - I used it for debugging, and may again, but I don't expect myself or anyone else to use it, and it may get yanked down the line.